### US Census Bureau 2017 Median Income by Census Tract ID

In [359]:
import pandas as pd
import numpy as np
import pickle

In [360]:
df1 = pd.read_csv('acs2017_census_tract_data.csv')

In [361]:
df1.head()

TractId    State          County  TotalPop   Men  Women  Hispanic  \
0  1001020100  Alabama  Autauga County      1845   899    946       2.4   
1  1001020200  Alabama  Autauga County      2172  1167   1005       1.1   
2  1001020300  Alabama  Autauga County      3385  1533   1852       8.0   
3  1001020400  Alabama  Autauga County      4267  2001   2266       9.6   
4  1001020500  Alabama  Autauga County      9965  5054   4911       0.9   

   White  Black  Native  ...  Walk  OtherTransp  WorkAtHome  MeanCommute  \
0   86.3    5.2     0.0  ...   0.5          0.0         2.1         24.5   
1   41.6   54.5     0.0  ...   0.0          0.5         0.0         22.2   
2   61.4   26.5     0.6  ...   1.0          0.8         1.5         23.1   
3   80.3    7.1     0.5  ...   1.5          2.9         2.1         25.9   
4   77.5   16.4     0.0  ...   0.8          0.3         0.7         21.0   

   Employed  PrivateWork  PublicWork  SelfEmployed  FamilyWork  Unemployment  
0       881         74.2        21.2           4.5         0.0           4.6  
1       852         75.9        15.0           9.0         0.0           3.4  
2      1482         73.3        21.1           4.8         0.7           4.7  
3      1849         75.8        19.7           4.5         0.0           6.1  
4      4787         71.4        24.1           4.5         0.0           2.3  

[5 rows x 37 columns]

In [362]:
df1.columns

Index(['TractId', 'State', 'County', 'TotalPop', 'Men', 'Women', 'Hispanic',
       'White', 'Black', 'Native', 'Asian', 'Pacific', 'VotingAgeCitizen',
       'Income', 'IncomeErr', 'IncomePerCap', 'IncomePerCapErr', 'Poverty',
       'ChildPoverty', 'Professional', 'Service', 'Office', 'Construction',
       'Production', 'Drive', 'Carpool', 'Transit', 'Walk', 'OtherTransp',
       'WorkAtHome', 'MeanCommute', 'Employed', 'PrivateWork', 'PublicWork',
       'SelfEmployed', 'FamilyWork', 'Unemployment'],
      dtype='object')

In [363]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 74001 entries, 0 to 74000
Data columns (total 37 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   TractId           74001 non-null  int64  
 1   State             74001 non-null  object 
 2   County            74001 non-null  object 
 3   TotalPop          74001 non-null  int64  
 4   Men               74001 non-null  int64  
 5   Women             74001 non-null  int64  
 6   Hispanic          73305 non-null  float64
 7   White             73305 non-null  float64
 8   Black             73305 non-null  float64
 9   Native            73305 non-null  float64
 10  Asian             73305 non-null  float64
 11  Pacific           73305 non-null  float64
 12  VotingAgeCitizen  74001 non-null  int64  
 13  Income            72885 non-null  float64
 14  IncomeErr         72885 non-null  float64
 15  IncomePerCap      73256 non-null  float64
 16  IncomePerCapErr   73256 non-null  float6

In [364]:
df1 = df1[['TractId','State','County','TotalPop','Men','Women','Income']]

In [365]:
df1 = df1[df1.State == 'New York']

In [366]:
df1

TractId     State         County  TotalPop   Men  Women   Income
43195  36001000100  New York  Albany County      1939   846   1093  27250.0
43196  36001000200  New York  Albany County      4731  2014   2717  24091.0
43197  36001000300  New York  Albany County      5558  2515   3043  33454.0
43198  36001000401  New York  Albany County      2459   955   1504  74255.0
43199  36001000403  New York  Albany County      4695  2010   2685  70815.0
...            ...       ...            ...       ...   ...    ...      ...
48108  36123150100  New York   Yates County      6229  3075   3154  48303.0
48109  36123150200  New York   Yates County      5434  2497   2937  43684.0
48110  36123150300  New York   Yates County      5292  2609   2683  52285.0
48111  36123150400  New York   Yates County      3662  1667   1995  68452.0
48112  36123150500  New York   Yates County      4466  2221   2245  47468.0

[4918 rows x 7 columns]

In [367]:
counties = ['New York County', 'Kings County', 'Bronx County', 'Richmond County', 'Queens County']

In [368]:
df1 = df1[df1.County.isin(counties)]

In [369]:
df1

TractId     State           County  TotalPop   Men  Women   Income
43283  36005000100  New York     Bronx County      7411  6857    554      NaN
43284  36005000200  New York     Bronx County      5058  2502   2556  62773.0
43285  36005000400  New York     Bronx County      5944  2984   2960  82697.0
43286  36005001600  New York     Bronx County      6115  2351   3764  33118.0
43287  36005001900  New York     Bronx County      2817  1327   1490  40117.0
...            ...       ...              ...       ...   ...    ...      ...
47144  36085030302  New York  Richmond County      6560  3459   3101  76417.0
47145  36085031901  New York  Richmond County      2737  1060   1677   9053.0
47146  36085031902  New York  Richmond County      5140  2409   2731  36932.0
47147  36085032300  New York  Richmond County      1136   578    558  69239.0
47148  36085990100  New York  Richmond County         0     0      0      NaN

[2167 rows x 7 columns]

In [370]:
df1[df1.TractId == 36005004700]

Empty DataFrame
Columns: [TractId, State, County, TotalPop, Men, Women, Income]
Index: []

In [371]:
df1.County.unique()

array(['Bronx County', 'Kings County', 'New York County', 'Queens County',
       'Richmond County'], dtype=object)

In [372]:
df1.sort_values(by='Income',ascending=False,inplace=True)

<ipython-input-372-1310493e3be9>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df1.sort_values(by='Income',ascending=False,inplace=True)


In [373]:
df1.reset_index(drop=True,inplace=True)

In [374]:
df1.describe()

TractId      TotalPop           Men         Women         Income
count  2.167000e+03   2167.000000   2167.000000   2167.000000    2101.000000
mean   3.605476e+10   3950.194739   1882.744347   2067.450392   63874.756307
std    2.594875e+07   2276.690218   1076.582663   1236.212918   31130.758337
min    3.600500e+10      0.000000      0.000000      0.000000    9053.000000
25%    3.604703e+10   2406.000000   1143.000000   1236.000000   41898.000000
50%    3.604712e+10   3571.000000   1712.000000   1856.000000   59529.000000
75%    3.608103e+10   5058.000000   2403.000000   2632.500000   78906.000000
max    3.608599e+10  28937.000000  12818.000000  16119.000000  233646.000000

In [375]:
df1['Income_percentile'] = df1.Income.rank(pct=True)

<ipython-input-375-93efa279f867>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df1['Income_percentile'] = df1.Income.rank(pct=True)


In [376]:
df1

TractId     State           County  TotalPop   Men  Women    Income  \
0     36061011401  New York  New York County      1224   620    604  233646.0   
1     36047002100  New York     Kings County      4910  2514   2396  212159.0   
2     36061014200  New York  New York County      4583  1916   2667  208024.0   
3     36061015002  New York  New York County      5119  2040   3079  205145.0   
4     36061003100  New York  New York County      2720  1263   1457  202404.0   
...           ...       ...              ...       ...   ...    ...       ...   
2162  36081138502  New York    Queens County        37    37      0       NaN   
2163  36081990100  New York    Queens County         0     0      0       NaN   
2164  36085015400  New York  Richmond County         0     0      0       NaN   
2165  36085022800  New York  Richmond County        27     0     27       NaN   
2166  36085990100  New York  Richmond County         0     0      0       NaN   

      Income_percentile  
0              1.000000  
1              0.999524  
2              0.999048  
3              0.998572  
4              0.998096  
...                 ...  
2162                NaN  
2163                NaN  
2164                NaN  
2165                NaN  
2166                NaN  

[2167 rows x 8 columns]

In [377]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2167 entries, 0 to 2166
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   TractId            2167 non-null   int64  
 1   State              2167 non-null   object 
 2   County             2167 non-null   object 
 3   TotalPop           2167 non-null   int64  
 4   Men                2167 non-null   int64  
 5   Women              2167 non-null   int64  
 6   Income             2101 non-null   float64
 7   Income_percentile  2101 non-null   float64
dtypes: float64(2), int64(4), object(2)
memory usage: 135.6+ KB


*Note: "Income" column has 2101 values, missing 66 values (or approx 3%)*

In [378]:
df1.Income.quantile(0.5)    # income where 50% are below, matches the 50% number above df1.describe()

59529.0

In [379]:
df1['Score'] = (10*df1['Income_percentile']).apply(np.ceil)

<ipython-input-379-c4cf2e078872>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df1['Score'] = (10*df1['Income_percentile']).apply(np.ceil)


In [380]:
df1

TractId     State           County  TotalPop   Men  Women    Income  \
0     36061011401  New York  New York County      1224   620    604  233646.0   
1     36047002100  New York     Kings County      4910  2514   2396  212159.0   
2     36061014200  New York  New York County      4583  1916   2667  208024.0   
3     36061015002  New York  New York County      5119  2040   3079  205145.0   
4     36061003100  New York  New York County      2720  1263   1457  202404.0   
...           ...       ...              ...       ...   ...    ...       ...   
2162  36081138502  New York    Queens County        37    37      0       NaN   
2163  36081990100  New York    Queens County         0     0      0       NaN   
2164  36085015400  New York  Richmond County         0     0      0       NaN   
2165  36085022800  New York  Richmond County        27     0     27       NaN   
2166  36085990100  New York  Richmond County         0     0      0       NaN   

      Income_percentile  Score  
0              1.000000   10.0  
1              0.999524   10.0  
2              0.999048   10.0  
3              0.998572   10.0  
4              0.998096   10.0  
...                 ...    ...  
2162                NaN    NaN  
2163                NaN    NaN  
2164                NaN    NaN  
2165                NaN    NaN  
2166                NaN    NaN  

[2167 rows x 9 columns]

In [381]:
df1.Score.unique()

array([10.,  9.,  8.,  7.,  6.,  5.,  4.,  3.,  2.,  1., nan])

In [382]:
df1.head()

TractId     State           County  TotalPop   Men  Women    Income  \
0  36061011401  New York  New York County      1224   620    604  233646.0   
1  36047002100  New York     Kings County      4910  2514   2396  212159.0   
2  36061014200  New York  New York County      4583  1916   2667  208024.0   
3  36061015002  New York  New York County      5119  2040   3079  205145.0   
4  36061003100  New York  New York County      2720  1263   1457  202404.0   

   Income_percentile  Score  
0           1.000000   10.0  
1           0.999524   10.0  
2           0.999048   10.0  
3           0.998572   10.0  
4           0.998096   10.0

In [383]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2167 entries, 0 to 2166
Data columns (total 9 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   TractId            2167 non-null   int64  
 1   State              2167 non-null   object 
 2   County             2167 non-null   object 
 3   TotalPop           2167 non-null   int64  
 4   Men                2167 non-null   int64  
 5   Women              2167 non-null   int64  
 6   Income             2101 non-null   float64
 7   Income_percentile  2101 non-null   float64
 8   Score              2101 non-null   float64
dtypes: float64(3), int64(4), object(2)
memory usage: 152.5+ KB


In [384]:
df1_top30 = df1.head(30)

In [385]:
df1_top30

TractId     State           County  TotalPop   Men  Women    Income  \
0   36061011401  New York  New York County      1224   620    604  233646.0   
1   36047002100  New York     Kings County      4910  2514   2396  212159.0   
2   36061014200  New York  New York County      4583  1916   2667  208024.0   
3   36061015002  New York  New York County      5119  2040   3079  205145.0   
4   36061003100  New York  New York County      2720  1263   1457  202404.0   
5   36061006900  New York  New York County      2568  1406   1162  198636.0   
6   36061003300  New York  New York County      5415  2692   2723  193750.0   
7   36061013000  New York  New York County      3399  1479   1920  190208.0   
8   36061014700  New York  New York County      3436  1670   1766  190167.0   
9   36061012000  New York  New York County      3335  1358   1977  187292.0   
10  36061005800  New York  New York County      3355  1869   1486  183988.0   
11  36061000900  New York  New York County      1736   937    799  180547.0   
12  36061008603  New York  New York County      4772  2318   2454  167607.0   
13  36061010200  New York  New York County       126    82     44  167500.0   
14  36061016700  New York  New York County      6400  2784   3616  165753.0   
15  36061012800  New York  New York County      5950  2769   3181  165049.0   
16  36061011202  New York  New York County       266    97    169  163750.0   
17  36061031704  New York  New York County     10241  5141   5100  162030.0   
18  36081074900  New York    Queens County      1590   781    809  161750.0   
19  36047000900  New York     Kings County      4410  2304   2106  161144.0   
20  36061010900  New York  New York County       179    93     86  160511.0   
21  36081056100  New York    Queens County        62    32     30  160250.0   
22  36061005501  New York  New York County      4709  2344   2365  160121.0   
23  36061005200  New York  New York County      3642  1920   1722  159821.0   
24  36061014500  New York  New York County      5858  2882   2976  159222.0   
25  36047006700  New York     Kings County      3941  1905   2036  158194.0   
26  36061003900  New York  New York County      5621  2553   3068  157652.0   
27  36047000700  New York     Kings County      4054  1986   2068  157534.0   
28  36061005700  New York  New York County      2766  1265   1501  157176.0   
29  36061012200  New York  New York County      2963  1271   1692  156827.0   

    Income_percentile  Score  
0            1.000000   10.0  
1            0.999524   10.0  
2            0.999048   10.0  
3            0.998572   10.0  
4            0.998096   10.0  
5            0.997620   10.0  
6            0.997144   10.0  
7            0.996668   10.0  
8            0.996192   10.0  
9            0.995716   10.0  
10           0.995240   10.0  
11           0.994764   10.0  
12           0.994288   10.0  
13           0.993812   10.0  
14           0.993337   10.0  
15           0.992861   10.0  
16           0.992385   10.0  
17           0.991909   10.0  
18           0.991433   10.0  
19           0.990957   10.0  
20           0.990481   10.0  
21           0.990005   10.0  
22           0.989529   10.0  
23           0.989053   10.0  
24           0.988577   10.0  
25           0.988101   10.0  
26           0.987625   10.0  
27           0.987149   10.0  
28           0.986673   10.0  
29           0.986197   10.0

In [386]:
df2 = df1[['TractId','Score']]

In [387]:
df2['TractId'] = df2['TractId'].astype(str)

<ipython-input-387-feb782ee5ec4>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df2['TractId'] = df2['TractId'].astype(str)


In [388]:
df2

TractId  Score
0     36061011401   10.0
1     36047002100   10.0
2     36061014200   10.0
3     36061015002   10.0
4     36061003100   10.0
...           ...    ...
2162  36081138502    NaN
2163  36081990100    NaN
2164  36085015400    NaN
2165  36085022800    NaN
2166  36085990100    NaN

[2167 rows x 2 columns]

In [389]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2167 entries, 0 to 2166
Data columns (total 2 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   TractId  2167 non-null   object 
 1   Score    2101 non-null   float64
dtypes: float64(1), object(1)
memory usage: 34.0+ KB


In [390]:
list1 = list(df1_top30.TractId)

In [391]:
list1

[36061011401,
 36047002100,
 36061014200,
 36061015002,
 36061003100,
 36061006900,
 36061003300,
 36061013000,
 36061014700,
 36061012000,
 36061005800,
 36061000900,
 36061008603,
 36061010200,
 36061016700,
 36061012800,
 36061011202,
 36061031704,
 36081074900,
 36047000900,
 36061010900,
 36081056100,
 36061005501,
 36061005200,
 36061014500,
 36047006700,
 36061003900,
 36047000700,
 36061005700,
 36061012200]

In [392]:
list2 = [str(x) for x in list1]

In [393]:
list2

['36061011401',
 '36047002100',
 '36061014200',
 '36061015002',
 '36061003100',
 '36061006900',
 '36061003300',
 '36061013000',
 '36061014700',
 '36061012000',
 '36061005800',
 '36061000900',
 '36061008603',
 '36061010200',
 '36061016700',
 '36061012800',
 '36061011202',
 '36061031704',
 '36081074900',
 '36047000900',
 '36061010900',
 '36081056100',
 '36061005501',
 '36061005200',
 '36061014500',
 '36047006700',
 '36061003900',
 '36047000700',
 '36061005700',
 '36061012200']

In [394]:
type(list2[0])

str

### NYC Subway Station Data

In [395]:
import pandas as pd

In [396]:
nyc_df = pd.read_csv('http://web.mta.info/developers/data/nyct/subway/Stations.csv')

In [397]:
nyc_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 496 entries, 0 to 495
Data columns (total 13 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Station ID             496 non-null    int64  
 1   Complex ID             496 non-null    int64  
 2   GTFS Stop ID           496 non-null    object 
 3   Division               496 non-null    object 
 4   Line                   496 non-null    object 
 5   Stop Name              496 non-null    object 
 6   Borough                496 non-null    object 
 7   Daytime Routes         496 non-null    object 
 8   Structure              496 non-null    object 
 9   GTFS Latitude          496 non-null    float64
 10  GTFS Longitude         496 non-null    float64
 11  North Direction Label  478 non-null    object 
 12  South Direction Label  481 non-null    object 
dtypes: float64(2), int64(2), object(9)
memory usage: 50.5+ KB


In [398]:
nyc_df.head(10)

Station ID  Complex ID GTFS Stop ID Division                 Line  \
0           1           1          R01      BMT              Astoria   
1           2           2          R03      BMT              Astoria   
2           3           3          R04      BMT              Astoria   
3           4           4          R05      BMT              Astoria   
4           5           5          R06      BMT              Astoria   
5           6           6          R08      BMT              Astoria   
6           7         613          R11      BMT              Astoria   
7           8           8          R13      BMT              Astoria   
8           9           9          R14      BMT  Broadway - Brighton   
9          10          10          R15      BMT  Broadway - Brighton   

                Stop Name Borough Daytime Routes Structure  GTFS Latitude  \
0  Astoria - Ditmars Blvd       Q            N W  Elevated      40.775036   
1            Astoria Blvd       Q            N W  Elevated      40.770258   
2                   30 Av       Q            N W  Elevated      40.766779   
3                Broadway       Q            N W  Elevated      40.761820   
4                   36 Av       Q            N W  Elevated      40.756804   
5                   39 Av       Q            N W  Elevated      40.752882   
6      Lexington Av/59 St       M          N W R    Subway      40.762660   
7              5 Av/59 St       M          N W R    Subway      40.764811   
8            57 St - 7 Av       M        N Q R W    Subway      40.764664   
9                   49 St       M          N R W    Subway      40.759901   

   GTFS Longitude   North Direction Label South Direction Label  
0      -73.912034                     NaN             Manhattan  
1      -73.917843            Ditmars Blvd             Manhattan  
2      -73.921479  Astoria - Ditmars Blvd             Manhattan  
3      -73.925508  Astoria - Ditmars Blvd             Manhattan  
4      -73.929575  Astoria - Ditmars Blvd             Manhattan  
5      -73.932755  Astoria - Ditmars Blvd             Manhattan  
6      -73.967258                  Queens   Downtown & Brooklyn  
7      -73.973347                  Queens   Downtown & Brooklyn  
8      -73.980658         Uptown & Queens   Downtown & Brooklyn  
9      -73.984139         Uptown & Queens   Downtown & Brooklyn

In [399]:
nyc_df[nyc_df["Stop Name"] == "86 St"]

Station ID  Complex ID GTFS Stop ID Division                Line  \
37           38          38          R44      BMT              4th Av   
78           79          79          N10      BMT           Sea Beach   
157         158         158          A20      IND  8th Av - Fulton St   
311         311         311          121      IRT      Broadway - 7Av   
397         397         397          626      IRT        Lexington Av   
473         476         476          Q04      IND           Second Av   

    Stop Name Borough Daytime Routes Structure  GTFS Latitude  GTFS Longitude  \
37      86 St      Bk              R    Subway      40.622687      -74.028398   
78      86 St      Bk              N  Open Cut      40.592721      -73.978230   
157     86 St       M            B C    Subway      40.785868      -73.968916   
311     86 St       M              1    Subway      40.788644      -73.976218   
397     86 St       M          4 5 6    Subway      40.779492      -73.955589   
473     86 St       M              Q    Subway      40.777891      -73.951787   

    North Direction Label South Direction Label  
37              Manhattan                 95 St  
78              Manhattan          Coney Island  
157    Uptown & The Bronx   Downtown & Brooklyn  
311    Uptown & The Bronx              Downtown  
397    Uptown & The Bronx   Downtown & Brooklyn  
473                Uptown   Downtown & Brooklyn

In [400]:
nyc_df2 = nyc_df.groupby(['Stop Name','Daytime Routes'])[['Structure']].count().sort_values(by='Structure',ascending=False)

In [401]:
nyc_df2['Structure'].unique()

array([1])

In [402]:
nyc_df2.reset_index(inplace=True)

In [403]:
nyc_df2[nyc_df2['Stop Name'] == '86 St']

Stop Name Daytime Routes  Structure
436     86 St              1          1
465     86 St          4 5 6          1
467     86 St            B C          1
492     86 St              R          1
493     86 St              Q          1
494     86 St              N          1

In [404]:
nyc_df2.groupby('Stop Name')[['Structure']].sum().sort_values(by='Structure',ascending=False).head(20)

Structure
Stop Name                            
86 St                               6
Canal St                            6
Fulton St                           5
23 St                               5
Times Sq - 42 St                    4
96 St                               4
145 St                              4
125 St                              4
72 St                               3
Atlantic Av - Barclays Ctr          3
Van Siclen Av                       3
Avenue U                            3
Bay Pkwy                            3
28 St                               3
18 Av                               3
7 Av                                3
Kings Hwy                           3
Broadway Jct                        3
Chambers St                         3
Church Av                           3

In [405]:
nyc_df.loc[nyc_df['Stop Name'].str.contains('Roose')]

Station ID  Complex ID GTFS Stop ID Division         Line  \
222         222         222          B06      IND      63rd St   
267         267         616          G14      IND  Queens Blvd   

                      Stop Name Borough Daytime Routes Structure  \
222            Roosevelt Island       M              F    Subway   
267  Jackson Hts - Roosevelt Av       Q        E F M R    Subway   

     GTFS Latitude  GTFS Longitude   North Direction Label  \
222      40.759145      -73.953260                  Queens   
267      40.746644      -73.891338  Forest Hills - Jamaica   

    South Direction Label  
222             Manhattan  
267             Manhattan

In [406]:
nyc_df.loc[nyc_df['Division'] == 'RIT']

Empty DataFrame
Columns: [Station ID, Complex ID, GTFS Stop ID, Division, Line, Stop Name, Borough, Daytime Routes, Structure, GTFS Latitude, GTFS Longitude, North Direction Label, South Direction Label]
Index: []

In [407]:
# note there are two Broadway stations, one in Queens, one in Brooklyn
nyc_df.loc[nyc_df['Stop Name'] == 'Broadway']

Station ID  Complex ID GTFS Stop ID Division       Line Stop Name  \
3             4           4          R05      BMT    Astoria  Broadway   
286         286         286          G30      IND  Crosstown  Broadway   

    Borough Daytime Routes Structure  GTFS Latitude  GTFS Longitude  \
3         Q            N W  Elevated      40.761820      -73.925508   
286      Bk              G    Subway      40.706092      -73.950308   

      North Direction Label South Direction Label  
3    Astoria - Ditmars Blvd             Manhattan  
286                  Queens             Church Av

In [408]:
list3 = list(nyc_df['Borough'].unique())
list3

['Q', 'M', 'Bk', 'Bx', 'SI']

In [409]:
list4 = ['Queens', 'Manhattan', 'Brooklyn', 'Bronx', 'StatenIsland']

In [410]:
dict1 = dict()

In [411]:
for i in range(len(list3)):
    key = list3[i]
    dict1[key] = list4[i]

In [412]:
dict1

{'Q': 'Queens',
 'M': 'Manhattan',
 'Bk': 'Brooklyn',
 'Bx': 'Bronx',
 'SI': 'StatenIsland'}

In [413]:
nyc_df['Borough2'] = nyc_df.Borough.map(dict1)

In [414]:
nyc_df.head()

Station ID  Complex ID GTFS Stop ID Division     Line  \
0           1           1          R01      BMT  Astoria   
1           2           2          R03      BMT  Astoria   
2           3           3          R04      BMT  Astoria   
3           4           4          R05      BMT  Astoria   
4           5           5          R06      BMT  Astoria   

                Stop Name Borough Daytime Routes Structure  GTFS Latitude  \
0  Astoria - Ditmars Blvd       Q            N W  Elevated      40.775036   
1            Astoria Blvd       Q            N W  Elevated      40.770258   
2                   30 Av       Q            N W  Elevated      40.766779   
3                Broadway       Q            N W  Elevated      40.761820   
4                   36 Av       Q            N W  Elevated      40.756804   

   GTFS Longitude   North Direction Label South Direction Label Borough2  
0      -73.912034                     NaN             Manhattan   Queens  
1      -73.917843            Ditmars Blvd             Manhattan   Queens  
2      -73.921479  Astoria - Ditmars Blvd             Manhattan   Queens  
3      -73.925508  Astoria - Ditmars Blvd             Manhattan   Queens  
4      -73.929575  Astoria - Ditmars Blvd             Manhattan   Queens

In [415]:
nyc_df['Station_Name'] = nyc_df['Stop Name'] + "_" + nyc_df['Borough2']

In [416]:
nyc_df.head(10)

Station ID  Complex ID GTFS Stop ID Division                 Line  \
0           1           1          R01      BMT              Astoria   
1           2           2          R03      BMT              Astoria   
2           3           3          R04      BMT              Astoria   
3           4           4          R05      BMT              Astoria   
4           5           5          R06      BMT              Astoria   
5           6           6          R08      BMT              Astoria   
6           7         613          R11      BMT              Astoria   
7           8           8          R13      BMT              Astoria   
8           9           9          R14      BMT  Broadway - Brighton   
9          10          10          R15      BMT  Broadway - Brighton   

                Stop Name Borough Daytime Routes Structure  GTFS Latitude  \
0  Astoria - Ditmars Blvd       Q            N W  Elevated      40.775036   
1            Astoria Blvd       Q            N W  Elevated      40.770258   
2                   30 Av       Q            N W  Elevated      40.766779   
3                Broadway       Q            N W  Elevated      40.761820   
4                   36 Av       Q            N W  Elevated      40.756804   
5                   39 Av       Q            N W  Elevated      40.752882   
6      Lexington Av/59 St       M          N W R    Subway      40.762660   
7              5 Av/59 St       M          N W R    Subway      40.764811   
8            57 St - 7 Av       M        N Q R W    Subway      40.764664   
9                   49 St       M          N R W    Subway      40.759901   

   GTFS Longitude   North Direction Label South Direction Label   Borough2  \
0      -73.912034                     NaN             Manhattan     Queens   
1      -73.917843            Ditmars Blvd             Manhattan     Queens   
2      -73.921479  Astoria - Ditmars Blvd             Manhattan     Queens   
3      -73.925508  Astoria - Ditmars Blvd             Manhattan     Queens   
4      -73.929575  Astoria - Ditmars Blvd             Manhattan     Queens   
5      -73.932755  Astoria - Ditmars Blvd             Manhattan     Queens   
6      -73.967258                  Queens   Downtown & Brooklyn  Manhattan   
7      -73.973347                  Queens   Downtown & Brooklyn  Manhattan   
8      -73.980658         Uptown & Queens   Downtown & Brooklyn  Manhattan   
9      -73.984139         Uptown & Queens   Downtown & Brooklyn  Manhattan   

                    Station_Name  
0  Astoria - Ditmars Blvd_Queens  
1            Astoria Blvd_Queens  
2                   30 Av_Queens  
3                Broadway_Queens  
4                   36 Av_Queens  
5                   39 Av_Queens  
6   Lexington Av/59 St_Manhattan  
7           5 Av/59 St_Manhattan  
8         57 St - 7 Av_Manhattan  
9                49 St_Manhattan

In [417]:
nyc_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 496 entries, 0 to 495
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Station ID             496 non-null    int64  
 1   Complex ID             496 non-null    int64  
 2   GTFS Stop ID           496 non-null    object 
 3   Division               496 non-null    object 
 4   Line                   496 non-null    object 
 5   Stop Name              496 non-null    object 
 6   Borough                496 non-null    object 
 7   Daytime Routes         496 non-null    object 
 8   Structure              496 non-null    object 
 9   GTFS Latitude          496 non-null    float64
 10  GTFS Longitude         496 non-null    float64
 11  North Direction Label  478 non-null    object 
 12  South Direction Label  481 non-null    object 
 13  Borough2               496 non-null    object 
 14  Station_Name           496 non-null    object 
dtypes: flo

In [418]:
len(nyc_df['Station_Name'].unique())

392

In [419]:
from collections import defaultdict

In [420]:
dict2 = defaultdict(int)

In [421]:
for x in list(nyc_df['Station_Name']):
    dict2[x] += 1

In [422]:
dict2

defaultdict(int,
            {'Astoria - Ditmars Blvd_Queens': 1,
             'Astoria Blvd_Queens': 1,
             '30 Av_Queens': 1,
             'Broadway_Queens': 1,
             '36 Av_Queens': 1,
             '39 Av_Queens': 1,
             'Lexington Av/59 St_Manhattan': 1,
             '5 Av/59 St_Manhattan': 1,
             '57 St - 7 Av_Manhattan': 1,
             '49 St_Manhattan': 1,
             'Times Sq - 42 St_Manhattan': 4,
             '34 St - Herald Sq_Manhattan': 2,
             '28 St_Manhattan': 3,
             '23 St_Manhattan': 5,
             '14 St - Union Sq_Manhattan': 2,
             '8 St - NYU_Manhattan': 1,
             'Prince St_Manhattan': 1,
             'Canal St_Manhattan': 6,
             'City Hall_Manhattan': 1,
             'Cortlandt St_Manhattan': 1,
             'Rector St_Manhattan': 2,
             'Whitehall St_Manhattan': 1,
             'Court St_Brooklyn': 1,
             'Jay St - MetroTech_Brooklyn': 2,
             'DeKalb Av_Bro

In [423]:
nyc_df.loc[nyc_df['Station_Name'] == 'Canal St_Manhattan']

Station ID  Complex ID GTFS Stop ID Division                Line  \
17           18         623          R23      BMT            Broadway   
18           19         623          Q01      BMT    Manhattan Bridge   
103         104         623          M20      BMT             Jamaica   
169         169         169          A34      IND  8th Av - Fulton St   
325         325         325          135      IRT      Broadway - 7Av   
410         410         623          639      IRT        Lexington Av   

    Stop Name Borough Daytime Routes Structure  GTFS Latitude  GTFS Longitude  \
17   Canal St       M            R W    Subway      40.719527      -74.001775   
18   Canal St       M            N Q    Subway      40.718383      -74.000460   
103  Canal St       M            J Z    Subway      40.718092      -73.999892   
169  Canal St       M          A C E    Subway      40.720824      -74.005229   
325  Canal St       M              1    Subway      40.722854      -74.006277   
410  Canal St       M              6    Subway      40.718803      -74.000193   

    North Direction Label South Direction Label   Borough2        Station_Name  
17        Uptown & Queens   Downtown & Brooklyn  Manhattan  Canal St_Manhattan  
18        Uptown & Queens   Downtown & Brooklyn  Manhattan  Canal St_Manhattan  
103              Brooklyn              Broad St  Manhattan  Canal St_Manhattan  
169       Uptown - Queens   Downtown & Brooklyn  Manhattan  Canal St_Manhattan  
325    Uptown & The Bronx              Downtown  Manhattan  Canal St_Manhattan  
410    Uptown & The Bronx              Downtown  Manhattan  Canal St_Manhattan

In [424]:
# duplicates but lat/long are approx the same, Division is different, Daytime Routes are different
nyc_df.loc[nyc_df['Station_Name'] == 'Times Sq - 42 St_Manhattan']

Station ID  Complex ID GTFS Stop ID Division                 Line  \
10           11         611          R16      BMT  Broadway - Brighton   
317         317         611          127      IRT       Broadway - 7Av   
468         467         611          725      IRT             Flushing   
469         468         611          902      IRT  Lexington - Shuttle   

            Stop Name Borough Daytime Routes Structure  GTFS Latitude  \
10   Times Sq - 42 St       M        N Q R W    Subway      40.754672   
317  Times Sq - 42 St       M          1 2 3    Subway      40.755290   
468  Times Sq - 42 St       M              7    Subway      40.755477   
469  Times Sq - 42 St       M              S    Subway      40.755983   

     GTFS Longitude North Direction Label South Direction Label   Borough2  \
10       -73.986754       Uptown & Queens   Downtown & Brooklyn  Manhattan   
317      -73.987495    Uptown & The Bronx   Downtown & Brooklyn  Manhattan   
468      -73.987691                Queens  34 St - Hudson Yards  Manhattan   
469      -73.986229                   NaN         Grand Central  Manhattan   

                   Station_Name  
10   Times Sq - 42 St_Manhattan  
317  Times Sq - 42 St_Manhattan  
468  Times Sq - 42 St_Manhattan  
469  Times Sq - 42 St_Manhattan

In [425]:
nyc_df['Station_Division'] = nyc_df['Stop Name'] + "_" + nyc_df['Division']

In [426]:
nyc_df.head()

Station ID  Complex ID GTFS Stop ID Division     Line  \
0           1           1          R01      BMT  Astoria   
1           2           2          R03      BMT  Astoria   
2           3           3          R04      BMT  Astoria   
3           4           4          R05      BMT  Astoria   
4           5           5          R06      BMT  Astoria   

                Stop Name Borough Daytime Routes Structure  GTFS Latitude  \
0  Astoria - Ditmars Blvd       Q            N W  Elevated      40.775036   
1            Astoria Blvd       Q            N W  Elevated      40.770258   
2                   30 Av       Q            N W  Elevated      40.766779   
3                Broadway       Q            N W  Elevated      40.761820   
4                   36 Av       Q            N W  Elevated      40.756804   

   GTFS Longitude   North Direction Label South Direction Label Borough2  \
0      -73.912034                     NaN             Manhattan   Queens   
1      -73.917843            Ditmars Blvd             Manhattan   Queens   
2      -73.921479  Astoria - Ditmars Blvd             Manhattan   Queens   
3      -73.925508  Astoria - Ditmars Blvd             Manhattan   Queens   
4      -73.929575  Astoria - Ditmars Blvd             Manhattan   Queens   

                    Station_Name            Station_Division  
0  Astoria - Ditmars Blvd_Queens  Astoria - Ditmars Blvd_BMT  
1            Astoria Blvd_Queens            Astoria Blvd_BMT  
2                   30 Av_Queens                   30 Av_BMT  
3                Broadway_Queens                Broadway_BMT  
4                   36 Av_Queens                   36 Av_BMT

In [427]:
nyc_subway = nyc_df.groupby(['Division','Stop Name','Borough','Station_Division','Borough2'])[['GTFS Latitude', 'GTFS Longitude']].mean()
nyc_subway.reset_index(inplace=True)

In [428]:
nyc_subway.head(10)

Division         Stop Name Borough      Station_Division   Borough2  \
0      BMT              1 Av       M              1 Av_BMT  Manhattan   
1      BMT            104 St       Q            104 St_BMT     Queens   
2      BMT            111 St       Q            111 St_BMT     Queens   
3      BMT            121 St       Q            121 St_BMT     Queens   
4      BMT  14 St - Union Sq       M  14 St - Union Sq_BMT  Manhattan   
5      BMT             18 Av      Bk             18 Av_BMT   Brooklyn   
6      BMT             20 Av      Bk             20 Av_BMT   Brooklyn   
7      BMT             23 St       M             23 St_BMT  Manhattan   
8      BMT             25 Av      Bk             25 Av_BMT   Brooklyn   
9      BMT             25 St      Bk             25 St_BMT   Brooklyn   

   GTFS Latitude  GTFS Longitude  
0      40.730953      -73.981628  
1      40.695178      -73.844330  
2      40.697418      -73.836345  
3      40.700492      -73.828294  
4      40.735736      -73.990568  
5      40.614312      -73.996075  
6      40.610983      -73.991597  
7      40.741303      -73.989344  
8      40.597704      -73.986829  
9      40.660397      -73.998091

In [429]:
nyc_subway.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 452 entries, 0 to 451
Data columns (total 7 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Division          452 non-null    object 
 1   Stop Name         452 non-null    object 
 2   Borough           452 non-null    object 
 3   Station_Division  452 non-null    object 
 4   Borough2          452 non-null    object 
 5   GTFS Latitude     452 non-null    float64
 6   GTFS Longitude    452 non-null    float64
dtypes: float64(2), object(5)
memory usage: 24.8+ KB


#### *Note: 452 subway stations compared to >2000 census tracts.  Therefore not all census tracts will have a subway station in them.*

In [430]:
len(nyc_subway)

452

In [431]:
nyc_subway['census_tract_id'] = ""

In [432]:
nyc_subway

Division         Stop Name Borough      Station_Division      Borough2  \
0        BMT              1 Av       M              1 Av_BMT     Manhattan   
1        BMT            104 St       Q            104 St_BMT        Queens   
2        BMT            111 St       Q            111 St_BMT        Queens   
3        BMT            121 St       Q            121 St_BMT        Queens   
4        BMT  14 St - Union Sq       M  14 St - Union Sq_BMT     Manhattan   
..       ...               ...     ...                   ...           ...   
447      SIR   Richmond Valley      SI   Richmond Valley_SIR  StatenIsland   
448      SIR         St George      SI         St George_SIR  StatenIsland   
449      SIR         Stapleton      SI         Stapleton_SIR  StatenIsland   
450      SIR     Tompkinsville      SI     Tompkinsville_SIR  StatenIsland   
451      SIR       Tottenville      SI       Tottenville_SIR  StatenIsland   

     GTFS Latitude  GTFS Longitude census_tract_id  
0        40.730953      -73.981628                  
1        40.695178      -73.844330                  
2        40.697418      -73.836345                  
3        40.700492      -73.828294                  
4        40.735736      -73.990568                  
..             ...             ...             ...  
447      40.519631      -74.229141                  
448      40.643748      -74.073643                  
449      40.627915      -74.075162                  
450      40.636949      -74.074835                  
451      40.512764      -74.251961                  

[452 rows x 8 columns]

In [433]:
import requests
import urllib

In [434]:
lat = 40.836488
lon = -73.847036

In [435]:
#Encode parameters
params = urllib.parse.urlencode({'latitude': lat, 'longitude':lon,'format':'json', 'showall':True})

In [436]:
params

'latitude=40.836488&longitude=-73.847036&format=json&showall=True'

In [437]:
#Construct request URL
url = 'https://geo.fcc.gov/api/census/block/find?' + params

In [438]:
url

'https://geo.fcc.gov/api/census/block/find?latitude=40.836488&longitude=-73.847036&format=json&showall=True'

In [439]:
response = requests.get(url)

In [440]:
response

<Response [200]>

In [441]:
#Parse json in response
data = response.json()

In [442]:
data
# Block = State + County + Tract + Block (# of digits: 2 + 3 + 6 + 4)

{'messages': ["FCC0001: The coordinate lies on the boundary of mulitple blocks, the block contains the clicked location is selected. For a complete list use showall=true to display 'intersection' element in the Block"],
 'Block': {'FIPS': '360050096002027',
  'bbox': [-73.848701, 40.834525, -73.845427, 40.836794]},
 'County': {'FIPS': '36005', 'name': 'Bronx'},
 'State': {'FIPS': '36', 'code': 'NY', 'name': 'New York'},
 'status': 'OK',
 'executionTime': '0'}

In [443]:
requests.get(url).json()['Block']['FIPS'][:11]

'36005009600'

In [444]:
for i in range(len(nyc_subway)):
    lat = nyc_subway.iloc[i,5]
    lon = nyc_subway.iloc[i,6]
    params = urllib.parse.urlencode({'latitude': lat, 'longitude':lon,'format':'json'})
    url = 'https://geo.fcc.gov/api/census/block/find?' + params
    nyc_subway.iloc[i,len(nyc_subway.columns)-1] = requests.get(url).json()['Block']['FIPS'][:11]

In [445]:
nyc_subway

Division         Stop Name Borough      Station_Division      Borough2  \
0        BMT              1 Av       M              1 Av_BMT     Manhattan   
1        BMT            104 St       Q            104 St_BMT        Queens   
2        BMT            111 St       Q            111 St_BMT        Queens   
3        BMT            121 St       Q            121 St_BMT        Queens   
4        BMT  14 St - Union Sq       M  14 St - Union Sq_BMT     Manhattan   
..       ...               ...     ...                   ...           ...   
447      SIR   Richmond Valley      SI   Richmond Valley_SIR  StatenIsland   
448      SIR         St George      SI         St George_SIR  StatenIsland   
449      SIR         Stapleton      SI         Stapleton_SIR  StatenIsland   
450      SIR     Tompkinsville      SI     Tompkinsville_SIR  StatenIsland   
451      SIR       Tottenville      SI       Tottenville_SIR  StatenIsland   

     GTFS Latitude  GTFS Longitude census_tract_id  
0        40.730953      -73.981628     36061003400  
1        40.695178      -73.844330     36081002600  
2        40.697418      -73.836345     36081012400  
3        40.700492      -73.828294     36081014400  
4        40.735736      -73.990568     36061005200  
..             ...             ...             ...  
447      40.519631      -74.229141     36085022600  
448      40.643748      -74.073643     36085000300  
449      40.627915      -74.075162     36085002100  
450      40.636949      -74.074835     36085002100  
451      40.512764      -74.251961     36085024800  

[452 rows x 8 columns]

In [446]:
nyc_subway.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 452 entries, 0 to 451
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Division          452 non-null    object 
 1   Stop Name         452 non-null    object 
 2   Borough           452 non-null    object 
 3   Station_Division  452 non-null    object 
 4   Borough2          452 non-null    object 
 5   GTFS Latitude     452 non-null    float64
 6   GTFS Longitude    452 non-null    float64
 7   census_tract_id   452 non-null    object 
dtypes: float64(2), object(6)
memory usage: 28.4+ KB


In [447]:
type(nyc_subway.iloc[0,3])

str

In [448]:
nyc_subway[nyc_subway['census_tract_id'].isin(list2)]

Division                    Stop Name Borough  \
4        BMT             14 St - Union Sq       M   
34       BMT                   8 St - NYU       M   
67       BMT                    City Hall       M   
72       BMT                     Court St      Bk   
125      BMT                 Whitehall St       M   
155      IND         34 St - Penn Station       M   
162      IND                   5 Av/53 St       M   
165      IND      59 St - Columbus Circle       M   
202      IND                     Canal St       M   
243      IND           Lexington Av/63 St       M   
272      IND                      York St      Bk   
302      IRT                        23 St       M   
306      IRT                        28 St       M   
320      IRT      59 St - Columbus Circle       M   
322      IRT       68 St - Hunter College       M   
326      IRT                        77 St       M   
327      IRT                        79 St       M   
333      IRT                     Astor Pl       M   
344      IRT  Brooklyn Bridge - City Hall       M   
370      IRT                  Franklin St       M   
412      IRT                  South Ferry       M   

                    Station_Division   Borough2  GTFS Latitude  \
4               14 St - Union Sq_BMT  Manhattan      40.735736   
34                    8 St - NYU_BMT  Manhattan      40.730328   
67                     City Hall_BMT  Manhattan      40.713282   
72                      Court St_BMT   Brooklyn      40.694100   
125                 Whitehall St_BMT  Manhattan      40.703087   
155         34 St - Penn Station_IND  Manhattan      40.752287   
162                   5 Av/53 St_IND  Manhattan      40.760167   
165      59 St - Columbus Circle_IND  Manhattan      40.768296   
202                     Canal St_IND  Manhattan      40.720824   
243           Lexington Av/63 St_IND  Manhattan      40.764629   
272                      York St_IND   Brooklyn      40.701397   
302                        23 St_IRT  Manhattan      40.741973   
306                        28 St_IRT  Manhattan      40.745143   
320      59 St - Columbus Circle_IRT  Manhattan      40.768247   
322       68 St - Hunter College_IRT  Manhattan      40.768141   
326                        77 St_IRT  Manhattan      40.773620   
327                        79 St_IRT  Manhattan      40.783934   
333                     Astor Pl_IRT  Manhattan      40.730054   
344  Brooklyn Bridge - City Hall_IRT  Manhattan      40.713065   
370                  Franklin St_IRT  Manhattan      40.719318   
412                  South Ferry_IRT  Manhattan      40.702068   

     GTFS Longitude census_tract_id  
4        -73.990568     36061005200  
34       -73.992629     36061005700  
67       -74.006978     36061003100  
72       -73.991777     36047000900  
125      -74.012994     36061000900  
155      -73.993391     36061010900  
162      -73.975224     36061010200  
165      -73.981736     36061014500  
202      -74.005229     36061003300  
243      -73.966113     36061012000  
272      -73.986751     36047002100  
302      -73.991128     36061005800  
306      -73.988814     36061005800  
320      -73.981929     36061014500  
322      -73.963870     36061012000  
326      -73.959874     36061012800  
327      -73.979917     36061016700  
333      -73.991070     36061005700  
344      -74.004131     36061003100  
370      -74.006886     36061003300  
412      -74.013664     36061000900

In [449]:
nyc_subway_final = nyc_subway.merge(df2, how='inner', left_on='census_tract_id', right_on='TractId')

In [450]:
nyc_subway_final

Division         Stop Name Borough      Station_Division      Borough2  \
0        BMT              1 Av       M              1 Av_BMT     Manhattan   
1        BMT            104 St       Q            104 St_BMT        Queens   
2        BMT            111 St       Q            111 St_BMT        Queens   
3        BMT            121 St       Q            121 St_BMT        Queens   
4        BMT  14 St - Union Sq       M  14 St - Union Sq_BMT     Manhattan   
..       ...               ...     ...                   ...           ...   
447      SIR   Richmond Valley      SI   Richmond Valley_SIR  StatenIsland   
448      SIR      Prince's Bay      SI      Prince's Bay_SIR  StatenIsland   
449      SIR         St George      SI         St George_SIR  StatenIsland   
450      SIR         Stapleton      SI         Stapleton_SIR  StatenIsland   
451      SIR     Tompkinsville      SI     Tompkinsville_SIR  StatenIsland   

     GTFS Latitude  GTFS Longitude census_tract_id      TractId  Score  
0        40.730953      -73.981628     36061003400  36061003400    8.0  
1        40.695178      -73.844330     36081002600  36081002600    8.0  
2        40.697418      -73.836345     36081012400  36081012400    7.0  
3        40.700492      -73.828294     36081014400  36081014400    4.0  
4        40.735736      -73.990568     36061005200  36061005200   10.0  
..             ...             ...             ...          ...    ...  
447      40.519631      -74.229141     36085022600  36085022600   10.0  
448      40.525507      -74.200064     36085019800  36085019800   10.0  
449      40.643748      -74.073643     36085000300  36085000300    7.0  
450      40.627915      -74.075162     36085002100  36085002100    5.0  
451      40.636949      -74.074835     36085002100  36085002100    5.0  

[452 rows x 10 columns]

In [451]:
nyc_subway_final.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 452 entries, 0 to 451
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Division          452 non-null    object 
 1   Stop Name         452 non-null    object 
 2   Borough           452 non-null    object 
 3   Station_Division  452 non-null    object 
 4   Borough2          452 non-null    object 
 5   GTFS Latitude     452 non-null    float64
 6   GTFS Longitude    452 non-null    float64
 7   census_tract_id   452 non-null    object 
 8   TractId           452 non-null    object 
 9   Score             433 non-null    float64
dtypes: float64(3), object(7)
memory usage: 38.8+ KB


*Note: 19 missing values in 'Score' column.*

In [452]:
nyc_subway_final.rename(columns={'Score':'HHIncomePct_score'},inplace=True)

In [453]:
del nyc_subway_final['TractId']

In [454]:
nyc_subway_final

Division         Stop Name Borough      Station_Division      Borough2  \
0        BMT              1 Av       M              1 Av_BMT     Manhattan   
1        BMT            104 St       Q            104 St_BMT        Queens   
2        BMT            111 St       Q            111 St_BMT        Queens   
3        BMT            121 St       Q            121 St_BMT        Queens   
4        BMT  14 St - Union Sq       M  14 St - Union Sq_BMT     Manhattan   
..       ...               ...     ...                   ...           ...   
447      SIR   Richmond Valley      SI   Richmond Valley_SIR  StatenIsland   
448      SIR      Prince's Bay      SI      Prince's Bay_SIR  StatenIsland   
449      SIR         St George      SI         St George_SIR  StatenIsland   
450      SIR         Stapleton      SI         Stapleton_SIR  StatenIsland   
451      SIR     Tompkinsville      SI     Tompkinsville_SIR  StatenIsland   

     GTFS Latitude  GTFS Longitude census_tract_id  HHIncomePct_score  
0        40.730953      -73.981628     36061003400                8.0  
1        40.695178      -73.844330     36081002600                8.0  
2        40.697418      -73.836345     36081012400                7.0  
3        40.700492      -73.828294     36081014400                4.0  
4        40.735736      -73.990568     36061005200               10.0  
..             ...             ...             ...                ...  
447      40.519631      -74.229141     36085022600               10.0  
448      40.525507      -74.200064     36085019800               10.0  
449      40.643748      -74.073643     36085000300                7.0  
450      40.627915      -74.075162     36085002100                5.0  
451      40.636949      -74.074835     36085002100                5.0  

[452 rows x 9 columns]

In [455]:
nyc_subway_final.loc[nyc_subway_final['Stop Name'] == '50 St']

Division Stop Name Borough Station_Division   Borough2  GTFS Latitude  \
30       BMT     50 St      Bk        50 St_BMT   Brooklyn      40.636260   
202      IND     50 St       M        50 St_IND  Manhattan      40.762456   
211      IRT     50 St       M        50 St_IRT  Manhattan      40.761728   

     GTFS Longitude census_tract_id  HHIncomePct_score  
30       -73.994791     36047022000                2.0  
202      -73.985984     36061013300                8.0  
211      -73.983849     36061013100               10.0

In [456]:
nyc_subway_final.loc[nyc_subway_final['Stop Name'] == 'Chambers St']

Division    Stop Name Borough Station_Division   Borough2  GTFS Latitude  \
84       BMT  Chambers St       M  Chambers St_BMT  Manhattan      40.713243   
245      IND  Chambers St       M  Chambers St_IND  Manhattan      40.714111   
247      IRT  Chambers St       M  Chambers St_IRT  Manhattan      40.715478   

     GTFS Longitude census_tract_id  HHIncomePct_score  
84       -74.003401     36061002900                1.0  
245      -74.008585     36061002100                NaN  
247      -74.009266     36061002100                NaN

In [457]:
list3 = nyc_subway_final['HHIncomePct_score'].unique()

In [458]:
sorted(list3)

[1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0, nan]

In [459]:
list_stations = sorted(nyc_subway_final['Station_Division'])

In [460]:
list_stations

['1 Av_BMT',
 '103 St - Corona Plaza_IRT',
 '103 St_IND',
 '103 St_IRT',
 '104 St_BMT',
 '104 St_IND',
 '110 St_IRT',
 '111 St_BMT',
 '111 St_IND',
 '111 St_IRT',
 '116 St - Columbia University_IRT',
 '116 St_IND',
 '116 St_IRT',
 '121 St_BMT',
 '125 St_IND',
 '125 St_IRT',
 '135 St_IND',
 '135 St_IRT',
 '137 St - City College_IRT',
 '138 St - Grand Concourse_IRT',
 '14 St - Union Sq_BMT',
 '14 St - Union Sq_IRT',
 '14 St_IND',
 '14 St_IRT',
 '145 St_IND',
 '145 St_IRT',
 '149 St - Grand Concourse_IRT',
 '15 St - Prospect Park_IND',
 '155 St_IND',
 '157 St_IRT',
 '161 St - Yankee Stadium_IND',
 '161 St - Yankee Stadium_IRT',
 '163 St - Amsterdam Av_IND',
 '167 St_IND',
 '167 St_IRT',
 '168 St - Washington Hts_IRT',
 '168 St_IND',
 '169 St_IND',
 '170 St_IND',
 '170 St_IRT',
 '174 St_IRT',
 '174-175 Sts_IND',
 '175 St_IND',
 '176 St_IRT',
 '18 Av_BMT',
 '18 Av_IND',
 '18 St_IRT',
 '181 St_IND',
 '181 St_IRT',
 '182-183 Sts_IND',
 '183 St_IRT',
 '190 St_IND',
 '191 St_IRT',
 '2 Av_IND',


In [461]:
with open('nyc_subway.pickle','wb') as to_write:
    pickle.dump(nyc_subway_final, to_write)

In [462]:
whos

Variable           Type              Data/Info
----------------------------------------------
counties           list              n=5
data               dict              n=6
defaultdict        type              <class 'collections.defaultdict'>
df1                DataFrame                   TractId     Sta<...>\n[2167 rows x 9 columns]
df1_top30          DataFrame                 TractId     State<...>        0.986197   10.0  
df2                DataFrame                   TractId  Score\<...>\n[2167 rows x 2 columns]
dict1              dict              n=5
dict2              defaultdict       defaultdict(<class 'int'><...>r Kill_StatenIsland': 1})
i                  int               451
key                str               SI
lat                float64           40.512764000000004
list1              list              n=30
list2              list              n=30
list3              ndarray           11: 11 elems, type `float64`, 88 bytes
list4              list              n=5
li

In [463]:
del nyc_subway_final

In [464]:
whos

Variable        Type              Data/Info
-------------------------------------------
counties        list              n=5
data            dict              n=6
defaultdict     type              <class 'collections.defaultdict'>
df1             DataFrame                   TractId     Sta<...>\n[2167 rows x 9 columns]
df1_top30       DataFrame                 TractId     State<...>        0.986197   10.0  
df2             DataFrame                   TractId  Score\<...>\n[2167 rows x 2 columns]
dict1           dict              n=5
dict2           defaultdict       defaultdict(<class 'int'><...>r Kill_StatenIsland': 1})
i               int               451
key             str               SI
lat             float64           40.512764000000004
list1           list              n=30
list2           list              n=30
list3           ndarray           11: 11 elems, type `float64`, 88 bytes
list4           list              n=5
list_stations   list              n=452
lon          

In [465]:
with open('nyc_subway.pickle','rb') as read_file:
    nyc_subway_final = pickle.load(read_file)

In [466]:
nyc_subway_final.head()

Division         Stop Name Borough      Station_Division   Borough2  \
0      BMT              1 Av       M              1 Av_BMT  Manhattan   
1      BMT            104 St       Q            104 St_BMT     Queens   
2      BMT            111 St       Q            111 St_BMT     Queens   
3      BMT            121 St       Q            121 St_BMT     Queens   
4      BMT  14 St - Union Sq       M  14 St - Union Sq_BMT  Manhattan   

   GTFS Latitude  GTFS Longitude census_tract_id  HHIncomePct_score  
0      40.730953      -73.981628     36061003400                8.0  
1      40.695178      -73.844330     36081002600                8.0  
2      40.697418      -73.836345     36081012400                7.0  
3      40.700492      -73.828294     36081014400                4.0  
4      40.735736      -73.990568     36061005200               10.0